internal/auth/oidc/rewrapping.go

```diff
@@ -0,0 +1,38 @@
+package oidc
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+)
+
+func init() {
+	kms.RegisterTableRewrapFn(defaultAuthMethodTableName, authMethodRewrapFn)
+}
+
+func authMethodRewrapFn(ctx context.Context, dataKeyVersionId, scopeId string, reader db.Reader, writer db.Writer, kmsRepo *kms.Kms) error {
+	const op = "oidc.authMethodRewrapFn"
+	var authMethods []*AuthMethod
+	// there are indexes on (scope id, <other>), so we can query on scope and refine via key id. this is the fastest query
+	if err := reader.SearchWhere(ctx, &authMethods, "scope_id=? and key_id=?", []interface{}{scopeId, dataKeyVersionId}, db.WithLimit(-1)); err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	wrapper, err := kmsRepo.GetWrapper(ctx, scopeId, kms.KeyPurposeDatabase)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	for _, am := range authMethods {
+		if err := am.decrypt(ctx, wrapper); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		if err := am.encrypt(ctx, wrapper); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		if _, err := writer.Update(ctx, am, []string{CtClientSecretField, ClientSecretHmacField, KeyIdField}, nil); err != nil {
```

Comment:
so I did make an assumption here that because the wrapper is passed into the hmac function (it's called cipher within the encrypt func), that the current encrypting key is used to calculate it, but I will double check this and get back to you regarding the consistency of calculated hmacs

Reply:
just confirmed, the [underlying hmac function](https://github.com/hashicorp/boundary/blob/main/internal/libs/crypto/hmac_sha256.go#L52-L56) and [bytes assigned to the reader](https://github.com/hashicorp/boundary/blob/main/internal/libs/crypto/derived_reader.go#L45) passed to the ed25519 key generator use the cipher bytes from the kms wrapper, so as long as the key version used to generate the hmac stays the same, the generated hmac will also stay the same.

-----
tested:
```
assert.NoError(t, got.decrypt(ctx, kmsWrapper2))
fmt.Printf("decrypted: %x\n", got.PasswordHmac)
assert.NoError(t, got.encrypt(ctx, kmsWrapper2))
fmt.Printf("encrypted: %x\n", got.PasswordHmac)
assert.NoError(t, got.decrypt(ctx, kmsWrapper2))
fmt.Printf("decrypted: %x\n", got.PasswordHmac)
assert.NoError(t, got.encrypt(ctx, kmsWrapper2))
fmt.Printf("encrypted: %x\n", got.PasswordHmac)
```
output
```
decrypted: 773b143a91bb516482497e8510c0a560c6fea6d7c146cc214166d2f9356ba5a5
encrypted: 773b143a91bb516482497e8510c0a560c6fea6d7c146cc214166d2f9356ba5a5
decrypted: 773b143a91bb516482497e8510c0a560c6fea6d7c146cc214166d2f9356ba5a5
encrypted: 773b143a91bb516482497e8510c0a560c6fea6d7c146cc214166d2f9356ba5a5
```

Reply:
so, to answer your question. we're updating the key version used to encrypt the fields, which updates the wrapper used by the hmac function, which updates key bytes passed to the ed25519 key generator. all of which results in a new hmac